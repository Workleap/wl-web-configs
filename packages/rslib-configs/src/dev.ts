import type { DistPathConfig, RsbuildPlugins, SourceMap } from "@rsbuild/core";
import { pluginReact, type PluginReactOptions } from "@rsbuild/plugin-react";
import { pluginSvgr, type PluginSvgrOptions } from "@rsbuild/plugin-svgr";
import { defineConfig, type Dts, type RsbuildConfigEntry, type RsbuildConfigOutputTarget, type Syntax } from "@rslib/core";
import { applyTransformers, type RslibConfigTransformer } from "./applyTransformers.ts";

export type DefineDevDefineReactPluginConfigFunction = (defaultOptions: PluginReactOptions) => PluginReactOptions;
export type DefineDevSvgrPluginConfigFunction = (defaultOptions: PluginSvgrOptions) => PluginSvgrOptions;

export interface DefineDevConfigOptions {
    entry?: RsbuildConfigEntry;
    syntax?: Syntax;
    bundle?: boolean;
    tsconfigPath?: string;
    dts?: Dts;
    target?: RsbuildConfigOutputTarget;
    distPath?: DistPathConfig;
    plugins?: RsbuildPlugins;
    sourceMap?: boolean | SourceMap;
    react?: false | DefineDevDefineReactPluginConfigFunction;
    svgr? : false | DefineDevSvgrPluginConfigFunction;
    transformers?: RslibConfigTransformer[];
}

function defaultDefineReactPluginConfig(options: PluginReactOptions) {
    return options;
}

function defineSvgrPluginConfig(options: PluginSvgrOptions) {
    return options;
}

export function defineDevConfig(options: DefineDevConfigOptions = {}) {
    const {
        entry: entryValue,
        syntax = "esnext",
        bundle = false,
        tsconfigPath,
        dts = false,
        target = "web",
        distPath = {
            root: "./dist"
        },
        plugins = [],
        sourceMap = {
            js: "cheap-module-source-map",
            css: true
        },
        react = defaultDefineReactPluginConfig,
        svgr = defineSvgrPluginConfig,
        transformers = []
    } = options;

    if (!bundle && !tsconfigPath) {
        throw new Error("[rslib-configs] When the \"bundle\" option is \"false\", a \"tsconfigPath\" option must be provided. We recommend including a tsconfig.build.json file specifically for compiling the library project.");
    }

    let entry = entryValue;

    if (!entry) {
        entry = {
            index: bundle ? ["./src/index.ts", "./src/index.js"] : "./src/**"
        };
    }

    const config = defineConfig({
        mode: "development",
        lib: [{
            format: "esm",
            syntax,
            bundle,
            dts
        }],
        source: {
            entry,
            tsconfigPath
        },
        output: {
            target,
            distPath,
            cleanDistPath: true,
            minify: false,
            sourceMap
        },
        plugins: [
            react && pluginReact(react({})),
            svgr && pluginSvgr(svgr({
                svgrOptions: {
                    exportType: "named"
                }
            })),
            ...plugins
        ]
    });

    const transformedConfig = applyTransformers(config, transformers, {
        environment: "dev"
    });

    return defineConfig(transformedConfig);
}
