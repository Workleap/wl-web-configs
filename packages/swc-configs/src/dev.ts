import type { Config as SwcConfig } from "@swc/core";
import { applyTransformers, type SwcConfigTransformer } from "./applyTransformers.ts";

export interface DefineDevConfigOptions {
    fastRefresh?: boolean;
    parser?: "ecmascript" | "typescript";
    transformers?: SwcConfigTransformer[];
}

export function defineDevConfig(targets: Record<string, string>, options: DefineDevConfigOptions = {}) {
    const {
        fastRefresh = true,
        parser = "typescript",
        transformers = []
    } = options;

    const config: SwcConfig = {
        jsc: {
            parser: parser === "ecmascript"
                ? {
                    syntax: "ecmascript",
                    jsx: true
                }
                : {
                    syntax: "typescript",
                    tsx: true
                },
            transform: {
                react: {
                    // Use "react/jsx-runtime".
                    runtime: "automatic",
                    // Use the native "Object.assign()" instead of "_extends".
                    useBuiltins: true,
                    // Enable React experimental "fast-refresh" feature.
                    // Also need to install @pmmmwh/react-refresh-webpack-plugin.
                    refresh: fastRefresh
                }
            },
            // Import shims from an external module rather than inlining them in bundle files to greatly reduce the bundles size.
            // Requires to add "@swc/helpers" as a project dependency.
            externalHelpers: true
        },
        module: {
            // The output module resolution system that the code will be compiled for.
            type: "es6",
            // Preserve dynamic imports.
            ignoreDynamic: true
        },
        env: {
            // jsc.target is not provided because the provided targets takes precedence.
            targets
        }
    };

    const transformedConfig = applyTransformers(config, transformers, {
        environment: "dev"
    });

    return transformedConfig;
}
