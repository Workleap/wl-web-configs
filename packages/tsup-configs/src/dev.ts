import type { Options as TsupConfig } from "tsup";
import { applyTransformers, type TsupConfigTransformer } from "./applyTransformers.ts";

export interface DefineDevConfigOptions extends TsupConfig {
    transformers?: TsupConfigTransformer[];
}

export function defineDevConfig(options: DefineDevConfigOptions = {}) {
    const {
        transformers = [],
        ...rest
    } = options;

    const config: TsupConfig = {
        dts: true,
        watch: true,
        entry: [
            "./src",
            "!src/**/*.css",
            "!src/**/*.stories.ts(x)",
            "!src/**/*.test.ts(x)"
        ],
        outDir: "./dist",
        format: "esm",
        target: "esnext",
        platform: "browser",
        sourcemap: "inline",
        ...rest
    };

    const transformedConfig = applyTransformers(config, transformers, {
        environment: "dev"
    });

    return transformedConfig;
}
