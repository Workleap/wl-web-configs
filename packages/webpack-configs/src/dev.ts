import ReactRefreshWebpackPlugin from "@pmmmwh/react-refresh-webpack-plugin";
import type { ReactRefreshPluginOptions } from "@pmmmwh/react-refresh-webpack-plugin/types/lib/types.d.ts";
import type { Config as SwcConfig } from "@swc/core";
import HtmlWebpackPlugin from "html-webpack-plugin";
import { createRequire } from "node:module";
import path from "path";
import { fileURLToPath } from "url";
import type { Configuration as WebpackConfig } from "webpack";
import webpack from "webpack";
import { applyTransformers, type WebpackConfigTransformer } from "./transformers/applyTransformers.ts";
import { isObject } from "./utils.ts";

// Add the "devServer" prop to WebpackConfig typings.
import "webpack-dev-server";

// Aliases
const DefinePlugin = webpack.DefinePlugin;

// Using node:module.createRequire until
// https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Operators/import.meta/resolve
// is available
const require = createRequire(import.meta.url);

export function defineDevHtmlWebpackPluginConfig(options: HtmlWebpackPlugin.Options = {}): HtmlWebpackPlugin.Options {
    const {
        template = path.resolve("./public/index.html"),
        ...rest
    } = options;

    return {
        ...rest,
        template
    };
}

export function defineFastRefreshPluginConfig(options: ReactRefreshPluginOptions = {}) {
    return options;
}

export interface DefineDevConfigOptions {
    entry?: string;
    https?: NonNullable<WebpackConfig["devServer"]>["https"];
    host?: string;
    port?: number;
    cache?: boolean;
    cacheDirectory?: string;
    moduleRules?: NonNullable<WebpackConfig["module"]>["rules"];
    plugins?: WebpackConfig["plugins"];
    htmlWebpackPlugin?: false | HtmlWebpackPlugin.Options;
    fastRefresh?: boolean | ReactRefreshPluginOptions;
    cssModules?: boolean;
    // Only accepting string values because there are lot of issues with the DefinePlugin related to typing errors.
    // See https://github.com/webpack/webpack/issues/8641
    environmentVariables?: Record<string, string | undefined>;
    transformers?: WebpackConfigTransformer[];
}

function preflight(options: DefineDevConfigOptions) {
    if (!require.resolve("webpack-dev-server")) {
        throw new Error("[webpack-configs] To use the \"dev\" config, install https://www.npmjs.com/package/webpack-dev-server as a \"devDependency\".");
    }

    if (options.fastRefresh) {
        if (!require.resolve("@pmmmwh/react-refresh-webpack-plugin")) {
            throw new Error("[webpack-configs] To use Webpack Fast Refresh, install https://www.npmjs.com/package/@pmmmwh/react-refresh-webpack-plugin as a \"devDependency\".");
        }
    }
}

function trySetSwcFastRefresh(config: SwcConfig, enabled: boolean) {
    if (config?.jsc?.transform?.react) {
        config.jsc.transform.react.refresh = enabled;
    }

    return config;
}

export function defineDevConfig(swcConfig: SwcConfig, options: DefineDevConfigOptions = {}) {
    preflight(options);

    const {
        entry = path.resolve("./src/index.tsx"),
        https = false,
        host = "localhost",
        port = 8080,
        cache = true,
        cacheDirectory = path.resolve("node_modules/.cache/webpack"),
        moduleRules = [],
        plugins = [],
        htmlWebpackPlugin = defineDevHtmlWebpackPluginConfig(),
        fastRefresh = true,
        cssModules = false,
        environmentVariables,
        transformers = []
    } = options;

    const config: WebpackConfig = {
        mode: "development",
        target: "web",
        devtool: "eval-cheap-module-source-map",
        devServer: {
            // According to the Fast Refresh plugin documentation, hot should be "true" to enable Fast Refresh:
            // https://github.com/pmmmwh/react-refresh-webpack-plugin#usage.
            hot: true,
            https,
            host,
            port,
            historyApiFallback: true
        },
        entry,
        output: {
            // The trailing / is very important, otherwise paths will not be resolved correctly.
            publicPath: `${https ? "https" : "http"}://${host}:${port}/`
        },
        cache: cache !== false && {
            type: "filesystem",
            allowCollectingMemory: true,
            buildDependencies: {
                config: [fileURLToPath(import.meta.url)]
            },
            cacheDirectory: cacheDirectory
        },
        optimization: {
            // See: https://webpack.js.org/guides/build-performance/#avoid-extra-optimization-steps
            runtimeChunk: true,
            removeAvailableModules: false,
            removeEmptyChunks: false,
            splitChunks: false
        },
        module: {
            rules: [
                {
                    test: /\.(js|jsx|ts|tsx)/i,
                    exclude: /node_modules/,
                    loader: require.resolve("swc-loader"),
                    options: trySetSwcFastRefresh(swcConfig, fastRefresh !== false)
                },
                {
                    // https://stackoverflow.com/questions/69427025/programmatic-webpack-jest-esm-cant-resolve-module-without-js-file-exten
                    test: /\.js/i,
                    include: /node_modules/,
                    resolve: {
                        fullySpecified: false
                    }
                },
                {
                    test: /\.css/i,
                    use: [
                        { loader: require.resolve("style-loader") },
                        {
                            loader: require.resolve("css-loader"),
                            options: cssModules
                                ? {
                                    // Must match the number of loaders applied before this one.
                                    importLoaders: 1,
                                    modules: true
                                }
                                : undefined
                        },
                        { loader: require.resolve("postcss-loader") }
                    ]
                },
                {
                    test: /\.svg/i,
                    loader: require.resolve("@svgr/webpack")
                },
                {
                    test: /\.(png|jpe?g|gif)/i,
                    type: "asset/resource"
                },
                ...moduleRules
            ]
        },
        resolve: {
            extensions: [".js", ".jsx", ".ts", ".tsx", ".css"],
            alias: {
                // Fixes Module not found: Error: Can't resolve '@swc/helpers/_/_class_private_field_init'.
                // View https://github.com/vercel/next.js/pull/38174 for more information and https://github.com/vercel/next.js/issues/48593.
                "@swc/helpers": path.dirname(require.resolve("@swc/helpers/package.json"))
            }
        },
        plugins: [
            htmlWebpackPlugin !== false && new HtmlWebpackPlugin(htmlWebpackPlugin as HtmlWebpackPlugin.Options),
            new DefinePlugin({
                "process.env": JSON.stringify(environmentVariables)
            }),
            fastRefresh && new ReactRefreshWebpackPlugin(isObject(fastRefresh) ? fastRefresh : defineFastRefreshPluginConfig()),
            ...plugins
        ].filter(Boolean) as WebpackConfig["plugins"]
    };

    const transformedConfig = applyTransformers(config, transformers, {
        environment: "dev"
    });

    return transformedConfig;
}
