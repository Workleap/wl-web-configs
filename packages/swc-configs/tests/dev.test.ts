import type { Config } from "@swc/core";
import { DefaultDevConfig, defineDevConfig } from "../src/dev.ts";
import type { ConfigOverride } from "../src/resolveOverrides.ts";

test("when no options are provided, return the default config", () => {
    const result = defineDevConfig();

    expect(result).toEqual(DefaultDevConfig);
});

test("when \"fastRefresh\" is true, the fast refresh configuration is included", () => {
    const result = defineDevConfig({
        fastRefresh: true
    });

    expect(result).toMatchSnapshot();
});

test("when \"parser\" is \"ecmascript\", the configuration parser is ecmascript", () => {
    const result = defineDevConfig({
        parser: "ecmascript"
    });

    expect(result).toMatchSnapshot();
});

test("when a config override function is provided, the function argument is the config with the non-config override options applied", () => {
    const expectedArgument = defineDevConfig({
        fastRefresh: true
    });

    const fct = jest.fn<Config, [Config]>(() => ({
        jsc: {
            parser: {
                syntax: "ecmascript"
            }
        }
    }));

    defineDevConfig({
        configOverride: fct as ConfigOverride
    });

    expect(fct).toHaveBeenCalledWith(expectedArgument);
});

test("providing options doesn't alter the default config object", () => {
    expect(DefaultDevConfig.jsc.parser.syntax).toBe("typescript");

    defineDevConfig({
        parser: "ecmascript"
    });

    expect(DefaultDevConfig.jsc.parser.syntax).toBe("typescript");
});

