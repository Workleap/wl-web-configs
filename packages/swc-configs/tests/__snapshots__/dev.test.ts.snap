// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`when "fastRefresh" is true, the fast refresh configuration is included 1`] = `
{
  "jsc": {
    "externalHelpers": true,
    "parser": {
      "syntax": "typescript",
      "tsx": true,
    },
    "target": "es2022",
    "transform": {
      "react": {
        "refresh": true,
        "runtime": "automatic",
        "useBuiltins": true,
      },
    },
  },
  "module": {
    "ignoreDynamic": true,
    "strict": true,
    "type": "es6",
  },
}
`;

exports[`when "parser" is "ecmascript", the configuration parser is ecmascript 1`] = `
{
  "jsc": {
    "externalHelpers": true,
    "parser": {
      "jsx": true,
      "syntax": "ecmascript",
    },
    "target": "es2022",
    "transform": {
      "react": {
        "refresh": true,
        "runtime": "automatic",
        "useBuiltins": true,
      },
    },
  },
  "module": {
    "ignoreDynamic": true,
    "strict": true,
    "type": "es6",
  },
}
`;
